signalr-go-server/pkg/signalr/hubconnection.go

```diff
@@ -0,0 +1,125 @@
+package signalr
+
+import (
+	"bytes"
+	"io"
+	"sync/atomic"
+)
+
+type hubConnection interface {
+	Start()
+	IsConnected() bool
+	Close(error string)
+	GetConnectionID() string
+	Receive() (interface{}, error)
+	SendInvocation(target string, args []interface{})
+	StreamItem(id string, item interface{})
+	Completion(id string, result interface{}, error string)
+	Ping()
+}
+
+type Connection interface {
+	io.Reader
+	io.Writer
+}
+
+func newHubConnection(connection Connection, connectionID string, protocol HubProtocol) hubConnection {
+	return &defaultHubConnection{
+		Protocol:     protocol,
+		ConnectionID: connectionID,
+		Writer:       connection,
+		Reader:       connection,
+	}
+}
+
+type defaultHubConnection struct {
+	Protocol     HubProtocol
+	ConnectionID string
+	Connected    int32
+	Writer       io.Writer
+	Reader       io.Reader
+}
+
+func (c *defaultHubConnection) Start() {
+	atomic.CompareAndSwapInt32(&c.Connected, 0, 1)
+}
+
+
+func (c *defaultHubConnection) IsConnected() bool {
+	return atomic.LoadInt32(&c.Connected) == 1
+}
+
+func (c *defaultHubConnection) Close(error string) {
+	atomic.StoreInt32(&c.Connected, 0)
+
+	var closeMessage = closeMessage{
+		Type:           7,
+		Error:          error,
+		AllowReconnect: true,
+	}
+	c.Protocol.WriteMessage(closeMessage, c.Writer)
+}
+
+func (c *defaultHubConnection) GetConnectionID() string {
+	return c.ConnectionID
+}
+
+func (c *defaultHubConnection) SendInvocation(target string, args []interface{}) {
+	var invocationMessage = sendOnlyHubInvocationMessage{
+		Type:      1,
+		Target:    target,
+		Arguments: args,
+	}
+
+	c.Protocol.WriteMessage(invocationMessage, c.Writer)
+}
+
+func (c *defaultHubConnection) Ping() {
+	var pingMessage = hubMessage{
+		Type: 6,
+	}
+
+	c.Protocol.WriteMessage(pingMessage, c.Writer)
+}
+
+func (c *defaultHubConnection) Receive() (interface{}, error) {
+	var buf bytes.Buffer
+	var data = make([]byte, 1<<15) // 32K
```

Comment:
Make the buffer 4K instead of 32. It's still inefficient but no performance work has been done as yet.